File: frontend/src/types/coverage-gap.ts
import type { GeoJSONFeature } from './api'
import type { GapSeverity, GapState } from './enums/gap-severity'
import type { SurveyArea } from './survey-area'

export interface CoverageGap {
  id:number; survey_area_id:number; source_run_ids:number[]; gap_geojson:GeoJSONFeature; area_square_m:number; gap_ratio:number; severity:GapSeverity
  recommended_line_geojson:GeoJSONFeature; algorithm_version:string; input_hash:string; gap_state:GapState; explanation:string; coverage_ratio:number
  overlap_ratio:number; processing_millis:number; version:number; detected_at:string; updated_at:string; survey_area?:SurveyArea
}
export interface CoverageEvidence { input_hash:string; coordinate_system:string; algorithm_version:string; source_run_count:number; coverage_ratio:number; overlap_ratio:number; gap_ratio:number; filtered_fragments:number; processing_millis:number; decision_boundary_note:string }
export interface CoverageResult { gap:CoverageGap; evidence:CoverageEvidence; idempotent:boolean }
export interface DetectCoverage { survey_area_id:number; source_run_ids:number[]; algorithm_version:string; resolution_m:number }

File: backend/internal/model/coverage_gap.go
package model

import (
	"time"

	"gorm.io/datatypes"
)

type CoverageGap struct {
	ID                     uint           `json:"id" gorm:"primaryKey"`
	SurveyAreaID           uint           `json:"survey_area_id" gorm:"not null;index"`
	SourceRunIDs           datatypes.JSON `json:"source_run_ids" gorm:"column:source_run_ids;type:jsonb;not null"`
	GapGeoJSON             datatypes.JSON `json:"gap_geojson" gorm:"column:gap_geojson;type:jsonb;not null"`
	AreaSquareM            float64        `json:"area_square_m" gorm:"not null"`
	GapRatio               float64        `json:"gap_ratio" gorm:"not null"`
	Severity               string         `json:"severity" gorm:"size:20;not null;index"`
	RecommendedLineGeoJSON datatypes.JSON `json:"recommended_line_geojson" gorm:"column:recommended_line_geojson;type:jsonb;not null"`
	AlgorithmVersion       string         `json:"algorithm_version" gorm:"size:40;not null"`
	InputHash              string         `json:"input_hash" gorm:"size:64;not null;uniqueIndex:idx_gap_idempotency"`
	GapState               string         `json:"gap_state" gorm:"size:24;not null;index"`
	Explanation            string         `json:"explanation" gorm:"size:1200;not null"`
	CoverageRatio          float64        `json:"coverage_ratio" gorm:"not null"`
	OverlapRatio           float64        `json:"overlap_ratio" gorm:"not null"`
	ProcessingMillis       int64          `json:"processing_millis" gorm:"not null"`
	Version                uint           `json:"version" gorm:"not null;default:1"`
	DetectedAt             time.Time      `json:"detected_at" gorm:"not null"`
	UpdatedAt              time.Time      `json:"updated_at"`
	SurveyArea             *SurveyArea    `json:"survey_area,omitempty" gorm:"foreignKey:SurveyAreaID;constraint:OnUpdate:CASCADE,OnDelete:RESTRICT"`
}

func (CoverageGap) TableName() string { return "coverage_gaps" }
